feat: add gorm dao support

chenjr15 opened the pull request.

chenjr15 committed:
feat: GORM-GEN 支持

--story=128281407

chenjr15 committed:
feat: 动态Structs支持更新结构

--story=128280650

chenjr15 committed:
feat: 泛型/动态Structs DAO支持

--story=128281023
--story=127880302

chenjr15 committed:
ci: add modernize linter

wcy00000000000000 commented on pkg/dal/orm/orm.go:
```diff
@@ -18,38 +18,331 @@
+	DB() *gorm.DB
+	DBContext(ctx context.Context) *gorm.DB
+	WriteDB(ctx context.Context) *gorm.DB
+	ReadDB(ctx context.Context) *gorm.DB
```
考虑把这几个获取db的方法统一成一个，通过参数控制，方便后续调整

wcy00000000000000 commented on pkg/structs/builder.go:
```diff
@@ -48,21 +51,27 @@ func GetBuilder(name string) (*Builder, bool) {
 	fieldIndexMap map[string]int
+	fields        []Field
+	// should not be used after invalid
+	invalid bool
```
bool类型的变量命名改成isInvalid这种吧，下面函数名也一起改一下吧

wcy00000000000000 commented on pkg/structs/builder.go:
```diff
@@ -48,21 +51,27 @@ func GetBuilder(name string) (*Builder, bool) {
 	validators map[string]func(any) error
 	// fieldIndexMap is the mapping of struct field name to their index in the struct.
 	fieldIndexMap map[string]int
+	fields        []Field
```
这个没有被用到

wcy00000000000000 commented on pkg/structs/builder.go:
```diff
@@ -103,7 +112,17 @@ func UpsertBuilderByFields(name string, fields []Field) (*Builder, error) {
+
+		if field.Anonymous && typ.Kind() == reflect.Struct {
+			for j := 0; j < typ.NumField(); j++ {
+				if typ.Field(i).Anonymous {
```
```suggestion
				if typ.Field(j).Anonymous {
```
对于匿名结构体的场景可以考虑把value改成数组记录一下它们的外层和内层的index，然后多层嵌套的场景现在是直接跳过了，是不是也应该记录一下

chenjr15 replied:
已修改为数组

wcy00000000000000 commented on pkg/structs/struct.go:
按现在的逻辑匿名结构体这里会重复调用

wcy00000000000000 commented on pkg/dal/types/page.go:
```diff
@@ -0,0 +1,236 @@
+)
+
+// PageWithoutSort page without sort.
+type PageWithoutSort struct {
```
为什么要单独定义一个没有sort的分页参数呢？

wcy00000000000000 commented on pkg/dal/types/page.go:
```diff
@@ -0,0 +1,236 @@
+	// DisableSort defines the sort field is not allowed to be defined by the user.
+	// then system defined sort field is used.
+	// Note: this option does not work when use the page to generate SQL expression,
+	// which means call the method of BasePage's SQLExprAndValue().
```
SQLExprAndValue这个是什么？

wcy00000000000000 commented on pkg/dal/types/page.go:
```diff
@@ -0,0 +1,236 @@
+	// Limit only works when the Count = false.
+	Limit uint64 `json:"limit"`
+	// Sort defines use which field to sort the queried resources.
+	// only 'one' field is supported to do sort.
```
如果想要支持用多个字段排序的话怎么办

chenjr15 replied:
改为数组了

wcy00000000000000 commented on pkg/dal/types/types.go:
```diff
@@ -0,0 +1,157 @@
+// CountOption defines options to count resources.
+type CountOption struct {
+	Filter  *filter.Expression
+	GroupBy string
```
GroupBy是不是应该放在单独的聚合请求参数里？

wcy00000000000000 commented on pkg/dal/types/types.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+// ValidateExcludeFilter validate list option, Filter is allowed to be empty.
+func (opt ListOption) ValidateExcludeFilter(eo *filter.ExprOption, po *PageOption) error {
```
filter是否允许为空可以通过参数控制而不是拆成两个函数，降低重复率

wcy00000000000000 commented on pkg/client/database/gorm.go:
```diff
@@ -0,0 +1,121 @@
+	var dialector gorm.Dialector
+	switch cfg.DBType {
+	case config.PostgreSQLType:
+		dsn := buildPostgreSQLDSN(cfg)
```
考虑后续使用config方式传参

chenjr15 replied:
拆分了DB连接创建方式为直接获取连接了

wcy00000000000000 commented on pkg/client/database/gorm.go:
```diff
@@ -0,0 +1,121 @@
+	default:
+		return nil, fmt.Errorf("unsupported db type: %s", cfg.DBType)
+	}
+	gormCfg := &gorm.Config{
```
这里只需要把使用的参数列出来，默认的没改的就不用了

wcy00000000000000 commented on pkg/config-center/config/db.go:
```diff
@@ -0,0 +1,64 @@
+
+const (
+	// PostgreSQLType ...
+	PostgreSQLType = "postgres"
```
这两个常量和下面DBConfig的DBType参数都应该是DBType类型吧

wcy00000000000000 commented on pkg/dal/table/table.go:
```diff
@@ -33,6 +35,29 @@ const (
+
+// RegisterBuilder register dynamic struct builder
+func RegisterBuilder(builder *structs.Builder) error {
+	tableRegistry[Name(builder.Name())] = builder
```
建议把动态结构单独抽一个map存储，动态和静态的处理逻辑不一样

wcy00000000000000 commented on pkg/dal/table/parser.go:
```diff
@@ -0,0 +1,202 @@
+	"github.com/TencentBlueKing/bk-cmdb/pkg/structs"
+)
+
+type tableFieldStore struct {
```
现在是直接用了里面的tableMap，理论上封装的话最好是把get和set方法一起封装了，不然为什么不直接用sync.Map呢？

wcy00000000000000 commented on pkg/dal/table/parser.go:
```diff
@@ -0,0 +1,202 @@
+
+// ModelTypeInfo store field type information of table fields
+type ModelTypeInfo struct {
+	fields       []*FieldInfo
```
数组和map的元素不需要用指针

wcy00000000000000 commented on pkg/dal/table/parser.go:
```diff
@@ -0,0 +1,202 @@
+func (mt *ModelTypeInfo) GetColumnTypeMap() map[string]filter.FieldType {
+	var colMap = make(map[string]filter.FieldType, len(mt.fields))
+	for i := 0; i < len(mt.fields); i++ {
+		colMap[mt.fields[i].Name] = mt.fields[i].ColumnType
```
用for range循环写起来更方便

wcy00000000000000 commented on pkg/dal/table/parser.go:
```diff
@@ -0,0 +1,202 @@
+func parseTableStruct(tableStruct any) ([]*FieldInfo, error) {
+	// ref: gorm.io/gen ConvertStructs
+	dummyDB, _ := gorm.Open(tests.DummyDialector{})
+	stmt := gorm.Statement{DB: dummyDB}
```
这个看起来有点奇怪，为什么要用一个测试db去parse？有没有别的方式直接解析的？

wcy00000000000000 commented on pkg/dal/orm/metric.go:
```diff
@@ -0,0 +1,62 @@
+	m := new(metric)
+	labels := prometheus.Labels{
+		// TODO 加上服务、环境信息，或者使用统一的进程ID
+		"connection_id": uuid.NewString(),
```
label一般不建议使用随机串，建议改成可以标识这个连接的信息

chenjr15 replied:
这个可能得和服务发现一起看下

wcy00000000000000 commented on pkg/dal/orm/orm.go:
```diff
@@ -18,38 +18,331 @@
+	}
+}
+
+// }
```
删掉

wcy00000000000000 commented on pkg/dal/dao/dao.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+// NewDao new dao instance
+func NewDao(kt context.Context, config *config.DBConfig) (Dao, error) {
```
context.Context命名用ctx吧，kit再用kt

wcy00000000000000 commented on pkg/dal/dao/dao.go:
```diff
@@ -0,0 +1,82 @@
+type dao struct {
+	orm orm.Interface
+	idg idgenerator.Interface
+	base.DynamicConstructor
```
这个为什么要用匿名参数？

chenjr15 replied:
为了实现Dynamic方法

wcy00000000000000 commented on pkg/dal/dao/dao.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+// TestModel ...
+func (d dao) TestModel() testmodel.Interface {
```
```suggestion
func (d *dao) TestModel() testmodel.Interface {
```

wcy00000000000000 commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,298 @@
+	modelName    string
+	modelBuilder *structs.Builder
+	Orm          orm.Interface
+	IDGen        idgenerator.Interface
```
这两个为什么要大写？理论上不会暴露出去吧

wcy00000000000000 commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,298 @@
+// BatchCreate batch create
+func (d *dynamicDao) BatchCreate(kt *kit.Kit, items *structs.Slice) (ids []string, err error) {
+	// 1. check builder
+	_, err = d.getBuilderBySlice(items)
```
这里只是校验类型的，实际上用的都是dao里的那个builder，命名调整一下吧
下面那个struct的同理

wcy00000000000000 commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,298 @@
+	}
+
+	return ids, nil
+
```
多余空行

wcy00000000000000 commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,298 @@
+}
+
+// refreshBuilder refresh model builder if invalid
+func (d *dynamicDao) refreshBuilder() error {
```
这个考虑做一个struct变更通知的逻辑，就不需要每次都调刷新接口了

wcy00000000000000 commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,298 @@
+
+// Update ...
+func (d *dynamicDao) Update(kt *kit.Kit, flt filter.RuleFactory, model *structs.Struct) (
+	updated int, err error) {
```
这个没超长不用换行，然后这个函数理论上也要校验一下model

wcy00000000000000 commented on pkg/dal/dao/base/generic.go:
```diff
@@ -0,0 +1,243 @@
+	"github.com/TencentBlueKing/bk-cmdb/pkg/log"
+)
+
+// IDSetterModel 对类型指针的约束
```
写一个详细的注释吧，然后现在的校验逻辑其实不生效，看一下怎么处理吧

ifooth commented on pkg/dal/dao/id-generator/id_generator_test.go:
```diff
@@ -292,7 +292,7 @@ func testParallelGen(t *testing.T, idGenFunc idGenFunc, ctx context.Context, res
-			idList, err := retryOnDuplicate(idGenFunc, ctx, resName, batchSize)
+			idList, err := retryOnDuplicate(idGenFunc, kt, resName, batchSize)
 			if err != nil {
 				t.Fatalf("fial to generate id list for %s, idx:%d, err: %v", resName, idx, err)
```
call to (*testing.T).Fatalf from a non-test goroutine

ifooth commented on pkg/config-center/config/db.go:
```diff
@@ -0,0 +1,64 @@
+
+	MaxIdleConns    int           `yaml:"max_idle_conns" json:"max_idle_conns"`
+	MaxOpenConns    int           `yaml:"max_open_conns" json:"max_open_conns"`
+	ConnMaxLifetime time.Duration `yaml:"conn_max_lifetime" json:"conn_max_lifetime"`
```
- 这个是内部使用的吧，配置文件的格式是int?， 这个要么建议就是统一秒，内部转换，或者str，内部解析，下同
- 没看到默认值（PS，这个内部约定大于配置，虽然也提供了配置，基本不改）

chenjr15 replied:
1. time.Duration 可以实现json/yaml解析，填写形如10s, 5m类似的duration字符串即可
2. 现在默认0是不按时间设置连接失效（默认值应该在配置文件中配置）

ifooth commented on pkg/config-center/config/db.go:
```diff
@@ -0,0 +1,64 @@
+	/* gorm settings */
+
+	// if > 0, all query will add this default timeout to context if given context timeout not specified
+	DefaultConnectionTimeout time.Duration `yaml:"default_connection_timeout" json:"default_connection_timeout"`
```
这个内部约定是否就OK

chenjr15 replied:
这里加的是未配置超时时间时的兜底默认约束，防止某些意外的超长时间的请求。

ifooth commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,298 @@
+	// BatchCreate batch create
+	BatchCreate(kt *kit.Kit, items *structs.Slice) (ids []string, err error)
+	List(kt *kit.Kit, opt *types.ListOption) (*types.DynamicListDetails, error)
+	Delete(kt *kit.Kit, filterExpr filter.RuleFactory) (updated int, err error)
```
updated int -> deleted int

wcy00000000000000 commented on pkg/dal/dao/base/generic.go:
```diff
@@ -0,0 +1,243 @@
+		return nil, fmt.Errorf("fail to get table fields for %s", tableName)
+	}
+	option := filter.NewExprOption(filter.RuleFields(fieldsInfo.GetColumnTypeMap()))
+	err := opt.Validate(option, types.NewDefaultPageOption())
```
这里直接用默认参数校验了，其他场景怎么办
然后filter理论上都应该是一套逻辑转换的，可以看看能不能抽

wcy00000000000000 commented on pkg/dal/dao/base/generic.go:
```diff
@@ -0,0 +1,243 @@
+	expr, err := conv.Filter(filterExpr)
+	if err != nil {
+		log.Error(kt, "fail to convert filter to clause expression",
+			log.E(err), slog.Any("filter", filterExpr))
```
没超长，下面也有一个一起改一下吧

ifooth commented on pkg/dal/gen/cmd/gen_main.go:
```diff
@@ -0,0 +1,56 @@
+	"github.com/TencentBlueKing/bk-cmdb/pkg/dal/table"
+)
+
+//go:generate go run gen_main.go -cwd ..
```
看有make gen, 但是又没有使用 //go:generate，这个要不去掉，要么保持go generate模式？

ifooth commented on pkg/dal/orm/orm.go:
```diff
@@ -18,38 +18,331 @@
+
+	ro := &runtimeOrm{db: db}
+	if ormOpts.ingressLimiter == nil {
+		ormOpts.ingressLimiter = rate.NewLimiter(rate.Limit(DefaultIngressLimit), DefaultIngressLimit)
```
几个问题
- 作为运维，不同环境db实例配置不一样，我应该填写哪个？
- 这里是实例级别的限流，如果负载不足，自动扩容了，会不会还是会把db打垮
- dba告警后，在线扩容了，配置是否还得跟着更新？

chenjr15 replied:
1. 不同环境，应该按对应环境配置即可
2. 会，实例级别的限流扩容后也可能导致DB被打垮
3. 需要更新

chenjr15 replied:
经讨论，暂不在orm中增加限流配置，统一在上层增加限流配置

Anna-shine commented on pkg/client/database/pgx.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+func buildPGXConfig(cfg *config.DBConfig, options ...PGXOption) (*pgx.ConnConfig, error) {
+	dsn := fmt.Sprintf("host=%s user=%s password=%s dbname=%s", cfg.Host, cfg.Username, cfg.Password, cfg.Database)
```
超长

Anna-shine commented on pkg/client/database/pgx.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+// OpenPGXStdlib construct a stdlib version pgx connection
+func OpenPGXStdlib(cfg *config.DBConfig, options ...PGXOption) (*sql.DB, error) {
```
这个命名看起来有点奇怪

chenjr15 replied:
PGX有两个版本接口，stdlib返回的是标准库版本的接口，是遵循PGX的命名习惯

Anna-shine commented on pkg/dal/dao/dao.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+// NewDao new dao instance
+func NewDao(ctx context.Context, config *config.DBConfig) (Dao, error) {
+	db, err := database.NewGORMClient(config)
```
一起确认下哪些位置需要打日志？
NewDao整个链路都没有打印日志

ifooth commented on pkg/dal/dao/base/dynamic_test.go:
```diff
@@ -0,0 +1,282 @@
+	})
+
+	// 注册表结构
+	table.RegisterBuilder(testDynamicBuilder)
```
这里没有这个函数

ifooth commented on pkg/dal/dao/base/generic.go:
```diff
@@ -0,0 +1,262 @@
+	}
+	for i := range items {
+		// type already checked above
+		any(&items[i]).(interface{ SetID(string) }).SetID(ids[i])
```
interface{ SetID(string) }  这个不是已经有定义了么
interface{ SetID(string) }).SetID(ids[i] -> any(&t).(IDSettable)

ifooth commented on pkg/dal/dao/base/generic.go:
```diff
@@ -0,0 +1,262 @@
+	if rt.Kind() == reflect.Ptr {
+		return fmt.Errorf("item types should not be ptr, got %T", t)
+	}
+	if _, ok := any(&t).(interface{ SetID(string) }); !ok {
```
同上

Anna-shine commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,281 @@
+	cfg := config{}
+	for _, opt := range opts {
+		opt(&cfg)
+	}
```
config没有被用到？

ifooth commented on pkg/client/database/pgx.go:
```diff
@@ -0,0 +1,74 @@
+type PGXOption func(connConfig *pgx.ConnConfig)
+
+// OpenPGX open pgx connection
+func OpenPGX(cfg *config.DBConfig, options ...PGXOption) (db *pgx.Conn, err error) {
```
返回pgx.Conn的貌似没有用到， 如果没有场景，可以先去掉，（OpenPGXStdlib -> OpenPGX, 和mysql的一致)

chenjr15 replied:
OpenPGX已去掉，但是不能去掉Stdlib，语义不一样

ifooth commented on pkg/client/database/mysql.go:
```diff
@@ -0,0 +1,68 @@
+	mysqlCfg := mysqldriver.NewConfig()
+	mysqlCfg.Addr = cfg.Host
+	if cfg.Port > 0 {
+		mysqlCfg.Addr = fmt.Sprintf("%s:%d", cfg.Host, cfg.Port)
```
拼接端口推荐 net.JoinHostPort()

ifooth commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,314 @@
+		return 0, fmt.Errorf("fail to convert filter to clause expression for update: %w", err)
+	}
+	// 2. check builder type
+	if !d.modelBuilder.Of(model) {
```
Of命名有点点奇怪,

ifooth commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,314 @@
+		return 0, fmt.Errorf("items is not a struct instance of model: %s", d.modelName)
+	}
+	ret := d.orm.DB(orm.WithContext(kt)).Table(d.modelName).Clauses(expr).Updates(model.Value())
+	return int(ret.RowsAffected), ret.Error
```
updated/deleted 类型和RowsAffected保持一致？

Anna-shine commented on pkg/dal/table/table.go:
```diff
@@ -33,27 +32,51 @@ const (
+
+// GetAllStaticTables get all static tables
+func GetAllStaticTables() []any {
+	var tables []any
```
```suggestion
	tables := make([]any, 0)
```

ifooth commented on pkg/dal/dao/base/generic.go:
```diff
@@ -0,0 +1,262 @@
+	}
+	err = d.Orm.DB().Create(items).Error
+	if err != nil {
+		log.Error(kt, "fail to create table", Table(tableName), log.E(err))
```
Table这个函数只是给log有点重了，如果是想默认加tabel参数，可以在struct初始化中自定义一个logger
```
logger := log.With("table", g.GetTableName())
d.logger = logger
return d
```
好处，不怕忘记 and 减少调用

ifooth commented on pkg/dal/dao/base/generic_test.go:
```diff
@@ -0,0 +1,147 @@
+	if rt.Kind() == reflect.Ptr {
+		panic("should not be ptr")
+	}
+	if _, ok := any(&t).(interface{ SetID(string) }); !ok {
```
interface 同 IDSettable

Anna-shine commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,314 @@
+	expr, eo, err := d.convFilter(kt, opt.Filter)
+	if err != nil {
+		log.Error(kt, "fail to convert filter to clause expression", "opt", opt, log.E(err))
+		return nil, fmt.Errorf("fail to convert filter to clause expression: %v", err)
```
%v

ifooth commented on pkg/structs/builder.go:
```diff
@@ -46,23 +59,51 @@ func GetBuilder(name string) (*Builder, bool) {
+// Listen for builder change.
+func Listen(handlerFunc func(name string, newBuilder *Builder)) uint64 {
+	builderLock.Lock()
+	handlerID++
```
这里的设计有待商榷，跨进程是否可用

wcy00000000000000 commented on pkg/dal/dao/base/config.go:
```diff
@@ -0,0 +1,43 @@
+	"github.com/TencentBlueKing/bk-cmdb/pkg/filter"
+)
+
+type config struct {
```
这个config配置对于一个dao的所有接口的校验都一样，考虑改成接口里支持配置可选的校验参数

Anna-shine commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,314 @@
+}
+
+// UpdateByID update by id
+func (d *dynamicDao) UpdateByID(kt *kit.Kit, id string, model *structs.Struct) (updated int, err error) {
```
注释错了

以及做个校验ret.RowsAffected数量？

Anna-shine commented on pkg/dal/dao/base/generic.go:
```diff
@@ -0,0 +1,262 @@
+	return slog.String("table", string(name))
+}
+
+// Generic 泛型通用DAO接口, T应为非指针类型, 且*T应该实现 IDSettable 接口,否则BatchCreate方法无法设置ID
```
这个接口可以挪到最上面

wcy00000000000000 commented on pkg/dal/types/page.go:
```diff
@@ -0,0 +1,215 @@
+	Sort []string `json:"sort"`
+	// Order is the direction when do sort operation.
+	// it works only when the Sort is set.
+	Order []Order `json:"order"`
```
这两个配套的参数建议组合成一个struct数组

Anna-shine commented on pkg/dal/types/types.go:
```diff
@@ -0,0 +1,112 @@
+	}
+	if err := opt.ValidateFields(eo); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (opt ListOption) validatePage(po *PageOption) error {
+	if opt.Page == nil {
+		return errors.New("page is required")
+	}
+
+	if po == nil {
+		return errors.New("page option is required")
+	}
+	if err := opt.Page.Validate(po); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ValidateFields check fields correctness, if fields is empty or rule fields is nil, skip check.
+func (opt ListOption) ValidateFields(eo *filter.ExprOption) error {
+	if eo.RuleFields == nil {
+		// skip check
+		return nil
+	}
+	if len(opt.Fields) == 0 {
+		return nil
+	}
+
+	for _, field := range opt.Fields {
+		if _, ok := eo.RuleFields[field]; !ok {
+			return fmt.Errorf("unknown field: %s", field)
+		}
+	}
+
+	return nil
+
+}
```
这里几个validate是ListOption几个参数的validate组合，实际在用的时候看List等还是用的裸的内部字段的validate，是不是重新考虑一下这个validate的设计

wcy00000000000000 commented on pkg/dal/table/parser.go:
```diff
@@ -0,0 +1,210 @@
+	"github.com/TencentBlueKing/bk-cmdb/pkg/structs"
+)
+
+type tableFieldStore struct {
```
考虑一下删除场景如何处理

wcy00000000000000 commented on pkg/dal/table/parser.go:
```diff
@@ -0,0 +1,210 @@
+}
+
+func (m *tableFieldStore) reloadTableStruct(ctx context.Context, tableName Name) (*ModelTypeInfo, bool) {
+	tableStruct, registered := GetStaticStruct(tableName)
```
static的可以初始化的时候都load进去，就不需要每次都判断了

wcy00000000000000 commented on pkg/structs/builder.go:
```diff
@@ -29,12 +30,24 @@ var (
 	builderLock sync.RWMutex
+
+	handlers  = make(map[uint64]Handler)
+	handlerID = uint64(0)
```
加注释

wcy00000000000000 commented on pkg/structs/builder.go:
```diff
@@ -29,12 +30,24 @@ var (
+		old.setInvalid()
+	}
 	builderRegistry[name] = b
+	for _, handler := range handlers {
+		handler(name, b)
+	}
```
放Unlock下面，不然handler里调用了获取builder方法的话会死锁

wcy00000000000000 commented on pkg/structs/builder.go:
```diff
@@ -46,23 +59,51 @@ func GetBuilder(name string) (*Builder, bool) {
+	builderLock.Lock()
+	handlerID++
+	id := handlerID
+	defer builderLock.Unlock()
```
这个位置怪怪的

wcy00000000000000 commented on pkg/structs/struct.go:
```diff
@@ -51,21 +52,32 @@ func (s *Struct) Get(fieldName string) (Valuer, error) {
 		return nil, fmt.Errorf("field %s is invalid", fieldName)
 	}
+
+	if field.Kind() == reflect.Struct && fieldName != field.Type().Name() &&
```
FieldByIndex我记得是会一层层拿的，这里为什么还要处理？

wcy00000000000000 commented on pkg/structs/struct.go:
```diff
@@ -81,6 +93,18 @@ func (s *Struct) Set(fieldName string, value any) error {
 		return nil
 	}
 
+	if field.Kind() == reflect.Struct && fieldName != field.Type().Name() &&
```
+1

wcy00000000000000 commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,299 @@
+	modelName string
+	orm       orm.Interface
+	idgen     idgenerator.Interface
+	Config
```
这个改成在参数里配置吧，不放在dao里了

chenjr15 replied:
目前是都支持的

wcy00000000000000 commented on pkg/dal/dao/test-model/test_model.go:
```diff
@@ -0,0 +1,63 @@
+}
+
+// GetNameByID Get name by id
+func (d dao) GetNameByID(kt *kit.Kit, id string) (string, error) {
```
```suggestion
func (d *dao) GetNameByID(kt *kit.Kit, id string) (string, error) {
```

wcy00000000000000 commented on pkg/dal/dao/base/dynamic.go:
```diff
@@ -0,0 +1,299 @@
+// Dynamic get new Dynamic dao with model
+func (d *dynamicConstructor) Dynamic(model string, opts ...Option) (Dynamic, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
```
先读，读不到再写，不要直接lock，或者注册个回调在增删动态struct的时候增删这个dao

chenjr15 replied:
不缓存了，直接用最新的

wcy00000000000000 commented on pkg/dal/dao/base/generic.go:
```diff
@@ -0,0 +1,268 @@
+	return ids, nil
+}
+
+func checkIDSettable[T any]() error {
```
考虑check放初始化

chenjr15 replied:
check 放初始化会污染整个调用链，所有构造都要加一个错误判断

chenjr15 committed:
feat: 新增独立的客户端structs支持嵌套和更新通知

- 数据库客户端解耦，新增独立的 MySQL/PostgreSQL 客户端
- ORM 接口改为函数式选项模式，移除旧的 DB 获取方法
- DAO 层增加排序和分页支持，优化类型处理
- 动态结构体支持多层嵌套和变更通知
- 移除ingress limit 限流功能

Breaking Change:
- ORM DB 获取方法签名变更
- ingress_limit 配置项不再生效

--story=128739720 db client 支持
--story=128750254 structs支持变更通知

wcy00000000000000 commented on pkg/dal/dao/base/config.go:
```diff
@@ -0,0 +1,57 @@
+// GetExprOption return filter.NewExprOption() merge with config.ExprOptionFunctions
+func GetExprOption(ruleFields map[string]filter.FieldType, opts []Option) *filter.ExprOption {
+	c := Config{
+		PageOption: types.NewDefaultPageOption(),
```
这个没用到

chenjr15 replied:
这个是用来保证option拿到的PageOption不会是空的

wcy00000000000000 commented on pkg/dal/dao/base/config.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+// GetPageOption return types.NewDefaultPageOption() if config.PageOption is unset
+func GetPageOption(opts []Option) *types.PageOption {
```
这个调用的地方都调了下面那个函数，可以直接一起转了，不用对config重复操作了

wcy00000000000000 commented on pkg/dal/table/registry.go:
```diff
@@ -0,0 +1,67 @@
+// Register for static table
+func Register(tableStruct types.Tabler) {
+	staticRegistry[types.Name(tableStruct.TableName())] = tableStruct
+	tableFields.loadStaticTableStruct()
```
每次register的时候都load一下所有的struct有点冗余，可以每次刷注册的这个或者最后统一刷一次

wcy00000000000000 commented on pkg/dal/table/registry.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+// GetStaticStruct get table struct by name
+func GetStaticStruct(name types.Name) (any, bool) {
```
这个目前没有用到，是预留给哪个场景的呢

ifooth commented on pkg/config-center/config/db.go:
```diff
@@ -0,0 +1,55 @@
+	MaxIdleConns       uint `yaml:"max_idle_conns" json:"max_idle_conns"`
+	MaxOpenConns       uint `yaml:"max_open_conns" json:"max_open_conns"`
+	ConnMaxLifeTimeSec uint `yaml:"conn_max_lifetime_sec" json:"conn_max_lifetime_sec,format:units"`
+	ConnMaxIdleTimeSec uint `yaml:"conn_max_idle_time_sec" json:"conn_max_idle_time_sec,format:units"`
```
conn_max_idle_time_sec下划线没一致?， sec已经有time的意思了，可以


conn_max_life_sec
conn_max_idle_sec
conn_max_life_seconds
conn_max_idle_seconds
上面几个我都OK

ifooth commented on pkg/dal/dao/base/generic.go:
```diff
@@ -0,0 +1,263 @@
+	tableName := d.GetTableName()
+	ids, err = d.IDGen.Batch(kt, tableName, uint64(len(items)))
+	if err != nil {
+		log.Error(kt, "fail to generate ids for table", "table", tableName, log.E(err))
```
日志保持一致？下同

chenjr15 committed:
feat: GORM 封装

--story=127880286
